les/enr_entry.go

```diff
@@ -34,10 +37,41 @@ func (e lesEntry) ENRKey() string {
 }
 
 // setupDiscovery creates the node discovery source for the eth protocol.
-func (eth *LightEthereum) setupDiscovery() (enode.Iterator, error) {
-	if len(eth.config.DiscoveryURLs) == 0 {
+func (eth *LightEthereum) setupDiscovery(cfg *p2p.Config) (enode.Iterator, error) {
+	var it enode.Iterator
+	if len(eth.config.DiscoveryURLs) != 0 {
+		client := dnsdisc.NewClient(dnsdisc.Config{})
+		var err error
+		it, err = client.NewIterator(eth.config.DiscoveryURLs...)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if cfg.DiscoveryV5 && eth.p2pServer.DiscV5 != nil {
+		v5 := eth.p2pServer.DiscV5.RandomNodes()
+		if it == nil {
+			it = v5
+		} else {
+			mix := enode.NewFairMix(0)
+			mix.AddSource(it)
+			mix.AddSource(v5)
+			it = mix
+		}
+	}
+	if it == nil {
 		return nil, nil
 	}
-	client := dnsdisc.NewClient(dnsdisc.Config{})
-	return client.NewIterator(eth.config.DiscoveryURLs...)
+	forkFilter := forkid.NewFilter(eth.blockchain)
+	return enode.Filter(it, func(n *enode.Node) bool {
```

Comment:
I think it's nice to filter the "non-les-server" here before establishing the TCP connection.
However, I don't think it's the proper place to check the forkID. It's better to check in the
handshake because we can check the nodes found by discovery and the nodes connect
us by itself(although I'm not sure it's realistic).

Reply:
This is a filter for discovery nodes. Checking forkID at handshake is a different story, you have an open PR to do that (and it was probably my mistake that it was delayed for so long). We can add that to the new (minimal) les4 upgrade.
Discv5 nodes are totally random and need to be filtered here because we don't want to TCP dial everyone. Nodes coming from DNS are pre-filtered by the crawler but it does not hurt to check them again here.